README.md
# Leaflet.Heightgraph

This plugin is inspired by [MrMufflon/Leaflet.Elevation](https://github.com/MrMufflon/Leaflet.Elevation).
You may use this plugin to view an interactive height profile of linestring segments using d3js.
The input data may consist of different types of attributes you wish to display.

![height_graph](https://user-images.githubusercontent.com/23240110/88930276-8efd5980-d27b-11ea-9b85-f6a864417f38.png)

Supported Browsers:
- Chrome
- Firefox
- Opera

[Demo](https://giscience.github.io/Leaflet.Heightgraph)

## Installation

Prerequisite:  [`node`](https://nodejs.org/en/download/) (version >=8)
or use [`nvm`](https://github.com/nvm-sh/nvm/blob/master/README.md) to install specific node versions

Install Leaflet.Heightgraph and dependencies with npm.
```
npm install leaflet.heightgraph
```

## Import

You can import the required libraries in the head of your index.html file
```html
 <link rel="stylesheet" href="node_modules/leaflet/dist/leaflet.css" />
 <script src="node_modules/leaflet/dist/leaflet.js"></script>
 <link rel="stylesheet" href="src/L.Control.Heightgraph.css"/>
 <script type="text/javascript" src="src/L.Control.Heightgraph.js"></script>
```

When using NPM you can require all needed libraries like this.
```
require('leaflet');
require('leaflet.heightgraph');
```

After importing Leaflet correctly, ES 6 Module import is possible as well:
```html
<script type="module">
    import 'leaflet.heightgraph'
</script>
```

The stylesheet can alternatively be imported in a style tag:
```html
<style>
    @import "../node_modules/leaflet.heightgraph/dist/L.Control.Heightgraph.min.css";
</style>
```

## Usage
Initialize the heightgraph, add it to your Leaflet map object and add your
Data to the heightgraph object.
```javascript
let hg = L.control.heightgraph();
hg.addTo(map);
hg.addData(geojson);
L.geoJson(geojson).addTo(map);
```

## Supported data
Input data has to be of type [GeoJSON-Format](http://geojson.org/).
This must consist of feature collection(s) corresponding to a certain
attribute which could be e.g. *surface* or *gradient* information.

Each `FeatureCollection` comprises a certain `attribute` in its `properties`
(e.g. `'summary': 'steepness'`) and has a list of `LineString` features.
These should have `coordinates` including height values and the `attributeType`
which corresponds to the certain type of attribute within this segment
(in this case it could be an index of steepness) declared in its `properties`.

Notice that the list of coordinates has to start with the last coordinate
of the previous `LineString`.

```javascript
const FeatureCollections = [{
    "type": "FeatureCollection",
    "features": [{
        "type": "Feature",
        "geometry": {
            "type": "LineString",
            "coordinates": [
                [8.6865264, 49.3859188, 114.5],
                [8.6864108, 49.3868472, 114.3],
                [8.6860538, 49.3903808, 114.8]
            ]
        },
        "properties": {
            "attributeType": "3"
        }
    }, {
        "type": "Feature",
        "geometry": {
            "type": "LineString",
            "coordinates": [
                [8.6860538, 49.3903808, 114.8],
                [8.6857921, 49.3936309, 114.4],
                [8.6860124, 49.3936431, 114.3]
            ]
        },
        "properties": {
            "attributeType": "0"
        }
    }],
    "properties": {
        "Creator": "OpenRouteService.org",
        "records": 2,
        "summary": "Steepness"
    }
}];
```

## Optional settings
These additional options can be set to customize your heightgraph.
Use them by passing an options object to the heightgraph during creation e.g.:
```js
let options = {
    position: "topleft"
}
let hg = L.control.heightgraph(options);
```

### position
You can choose between `"bottomright"`, `"bottomleft"`, `"topright"` and `"topleft"`
for the position on the map.

default: `position: "bottomright"`

### width
The width of the expanded heightgraph display in pixels.

default: `width: 800`

### height
The height of the expanded heightgraph display in pixels.

default: `height: 280`

### margins
The margins define the distance between the border of the heightgraph
and the actual graph inside. You are able to specify margins for `top`, 
`right`, `bottom` and `left` in pixels.

default:
```
margins: {
    top: 10,
    right: 30,
    bottom: 55,
    left: 50
}
```

### expand
Boolean value that defines if the heightgraph should be expanded on creation.

default: `true`

### expandControls
Boolean value that defines if the expand controls (i.e. toggle and close) should be visible.
To be set to false when the heightgraph is embedded in another component
and that controls when and how the heightgraph is expanded.

default: `true`

### expandCallback
Function to be called if the heightgraph is expanded or reduced. The state of
the heightgraph is passed as an argument. It is `true` when expanded and
`false` when reduced.

default: `undefined`

example:
```js
expandCallback: function(expanded){
    console.log("Expanded: " + expanded)
}
```

### mappings
You may add a mappings object to customize the colors and labels in the height graph.
Without adding custom mappings the segments and labels within the graph will be displayed in random colors.
Each key of the object must correspond to the `summary` key in `properties` within the `FeatureCollection`.

default: `undefined`

Example:

```javascript
colorMappings.Steepness = {
    '3': {
        text: '1-3%',
        color: '#F29898'
    },
    '0': {
        text: '4-6%',
        color: '#E07575'
    }
};
```

### highlightStyle
You can customize the highlight style when using the horizontal line to 
find parts of the route above an elevation value.
Use any [Leaflet Path options](https://leafletjs.com/reference-1.5.0.html#path-option)
as value of the `highlightStyle` parameter.

default: `highlightStyle:{color: 'red'}`

Example:
```javascript
highlightStyle = {
   weight: 10,
   opacity: 0.8,
   color: 'orange'
 }
```

### graphStyle
Allows customizing the style of the height graph.
You may specify the [<path> element presentation attributes](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/path) in the form of a JavaScript object.
Note that in case of conflict the `mappings` attributes take precedence over the `graphStyle` attributes.

default: `graphStyle: {}`

Example:
```javascript
graphStyle: {
    opacity: 0.8,
    'fill-opacity': 0.5,
    'stroke-width': '2px'
};
```

### translation
You can change the labels of the heightgraph info field by passing translations
for `distance`, `elevation`, `segment_length`, `type` and `legend`.

default:
```
translation: {
    distance: "Distance",
    elevation: "Elevation",
    segment_length: "Segment length",
    type: "Type",
    legend: "Legend"
    }
```

### xTicks
Overwrite automatic tick handling for x axis and specify the tick frequency in the *x axis* of the graph. Corresponds approximately to 
2 to the power of `value` ticks.

default: `xTicks: 3`

### yTicks
Overwrite automatic tick handling for y axis and specify the tick frequency in the *y axis* of the graph. Corresponds approximately to 
2 to the power of `value` ticks.

default: `yTicks: 3`

## Methods
The following methods are available on a created `L.control.heightgraph` instance,
represented by `hg` in the examples.

### mapMousemoveHandler
Used together with mapMouseoutHandler.
Takes a mousemove event as input `hg.mapMousemoveHandler(event)` to show
the current position on the heightgraph when hovering over the respective
geometry on the map.
The marker on the map can be disabled by passing an object with properties
`showMapMarker:false` as second argument. `showMapMarker` is `true` by default.
```js
hg.mapMousemoveHandler(mousemoveEvent, {showMapMarker: false})
```
Check [example.js](/example/example.js) for an implementation example.


### mapMouseoutHandler
Used together with mapMousemoveHandler.
Responsible for removing the created markers again.
You can pass a timeout in milliseconds to delay the removal.
The default timeout is `1000`.
```js
hg.mapMouseoutHandler(0)
```
Check [example.js](/example/example.js) for an implementation example.

### resize
Use this to resize the heightgraph container including the graph to the
specified extent by passing an object including defined `width` and `height`
properties in pixel values:
```js
hg.resize({width: 1000, height: 300})
```

## Development setup

```bash
# clone the repository
$ git clone https://github.com/GIScience/Leaflet.Heightgraph.git

# install dependencies using a node-version >= 8
$ npm install

# start development server
$ npm start
```

## Configurations (WebStorm)

You can create run configurations for different tasks:

### Starting development server

- open `Run -> Edit Configurations...`
- click `+` to add a new configuration and choose the npm template
- give the Configuration a name e.g. `Dev`
- choose `start` as command
- press `OK`

### Testing

Debug jasmine tests with karma in WebStorm

- open `Run -> Edit Configurations...`
- click `+` to add a new configuration and choose the karma template
- give the Configuration a name e.g. `Test`
- press `OK`

### Coverage

Run karma with coverage

- once you have a karma task configured just click the run with coverage button 
- analyse coverage in Webstorm or Browser (open ./coverage/html/index.html)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## v1.4.0 (2020-10-26)

**Added**
- Option pass a callback function in the mappings object. This can be used for path details that have many possible values. See the example mappings.js for an example ([#103](https://github.com/GIScience/Leaflet.Heightgraph/issues/103)).
- Add choose selection callback to remember what attribute index was selected ([#104](https://github.com/GIScience/Leaflet.Heightgraph/issues/104)).
- Add `selectedAttributeIdx` to options to allow choosing a different attribute index than 0 ([#105](https://github.com/GIScience/Leaflet.Heightgraph/issues/105)).

## v1.3.3 (2020-10-21)

**Added**
- human readable label in the data (see example/data.js) ([#99](https://github.com/GIScience/Leaflet.Heightgraph/issues/99))

**Fixed**
- undefined event ([#102](https://github.com/GIScience/Leaflet.Heightgraph/issues/102))

## v1.3.2 (2020-07-31)

**Changed**
- module field path in package.json

## v1.3.1 (2020-07-30)

**Fixed**
- using margin width and height in tick and grid generation ([#87](https://github.com/GIScience/Leaflet.Heightgraph/issues/87))

## v1.3.0 (2020-07-30)

**Changed**
- calculation of elevation bounds to a simpler method
- README.md height graph image

**Fixed**
- resize() resets chosen category ([#97](https://github.com/GIScience/Leaflet.Heightgraph/issues/97))
- overlapping axis ticks for small heightgraph size ([#87](https://github.com/GIScience/Leaflet.Heightgraph/issues/87))

## v1.2.0 (2020-07-25)

**Added**
- click option to legend to keep it open

**Fixed**
- hg.addData toggles expand/minimize ([#94](https://github.com/GIScience/Leaflet.Heightgraph/issues/94))
- resizing with less than 2 parameters crops height graph ([#93](https://github.com/GIScience/Leaflet.Heightgraph/issues/93))
- graphStyle not adjusting legend items([#89](https://github.com/GIScience/Leaflet.Heightgraph/issues/89))
- zooming out when changing category ([#90](https://github.com/GIScience/Leaflet.Heightgraph/issues/90))
- disappearing category selection arrows on resize ([#92](https://github.com/GIScience/Leaflet.Heightgraph/issues/92))

**Security**
- update dependencies
- `npm audit fix` lodash issues

## v1.1.0 (2020-07-05)

**Added**
- `graphStyle` option to customize the graph style ([PR #85](https://github.com/GIScience/Leaflet.Heightgraph/pull/85))
- `expandControls` option to hide the close button and expand by default ([PR #82](https://github.com/GIScience/Leaflet.Heightgraph/pull/82))
- `resize` method for dynamic resizing of the heightgraph ([#78](https://github.com/GIScience/Leaflet.Heightgraph/issues/78))
- `mapMousemoveHandler` method for handling a mousemove event to show the graph- and optionally the map-marker ([PR #82](https://github.com/GIScience/Leaflet.Heightgraph/pull/82))
- `mapMouseoutHandler` method for handling the mouseout event, removing the markers ([PR #82](https://github.com/GIScience/Leaflet.Heightgraph/pull/82))

**Changed**
- `addData` method functionality ([#37](https://github.com/GIScience/Leaflet.Heightgraph/issues/37)):
    - instead of removing and adding the whole height graph only the data
    is cleared and replaced, keeping the L.control.heightgraph instance
    - mappings for subsequent data sets need to be specified during
    initialization of the height graph
- dependencies to latest ([#83](https://github.com/GIScience/Leaflet.Heightgraph/issues/83))
- prefixed class names ([#65](https://github.com/GIScience/Leaflet.Heightgraph/issues/65))

**Removed**
- unused dependencies: bower, concurrently & connect-modrewrite

## v1.0.0 (2020-06-26)

**Added**
- rollup for building and bundling d3 ([#45](https://github.com/GIScience/Leaflet.Heightgraph/issues/45))
- modularization
- `es-dev-server` (replaces lite-server) for development

**Changed**
- folder structure

**Removed**
- removed d3 dependency
- grunt as task runner
- unused dev-dependencies
- lite-server

## v0.4.1 (2019-11-15)

**Added**
- `xTick` and `yTick` option to set axis tick frequency

## v0.4.0 (2019-11-14)

**Added**
- `highlightStyle` options parameter to customize the highlighting style
for the 'elevation above' feature (horizontal line)
- more documentation to README.md
- babel transpiler
([#70](https://github.com/GIScience/Leaflet.Heightgraph/issues/70))
- minified script versions and `img` folder to distribution folder

**Changed**
- highlighting geometry type from multiPolyline to multiple L.polylines
- default values for `height` and `margins` to the ones that are actually stated in the readme
- osm tile provider to `https` instead of `http`

**Fixed**
- security issues
([#66](https://github.com/GIScience/Leaflet.Heightgraph/issues/66))
- some console error where `this._pointG` was undefined

## v0.3.0 (2019-10-29)

**Added**
- Karma test runner
- Coverage report
- ES6 Support
- height selection on drag

**Fixed**
- option initialization on `addData()` call
- `_fitSection()` not working with round-trips
- index.html not working with undefined mappings
- horizontal line for height selection jumping on reselect
([#42](https://github.com/GIScience/Leaflet.Heightgraph/issues/42))

**Changed**
- version of `d3`, `grunt-contrib-jasmine` and `Leaflet` to latest
- eol to LF in all files
- dependency source from bower to node

## 0.2.0 (2019-08-23)

**API Changes**

* The class name `focus` was renamed to `focusbox`.
`focus` might be used in many other frameworks and could lead to css conflicts.
Please make sure you update the .css when updating the .js!
([#43](https://github.com/GIScience/Leaflet.Heightgraph/pull/64) by 
[boldtrn](https://github.com/boldtrn)).

## 0.1.3 (2019-07-17)

**Improvements**

* Allow customizing x and y axis ticks
([#61](https://github.com/GIScience/Leaflet.Heightgraph/pull/61) by
[boldtrn](https://github.com/boldtrn)).

## 0.1.2 (2019-07-17)

**Improvements**

* Callback for the expand state
([#58](https://github.com/GIScience/Leaflet.Heightgraph/pull/58) by

heightProfile.js
L.Control.Heightgraph = L.Control.extend({
    options: {
        position: "topleft",
        width: 800,
        height: 125,
        margins: {
            top: 45,
            right: 20,
            bottom: 30,
            left: 50
        }
    },
    onAdd: function(map) {
        var opts = this.options;
        var controlDiv = this._container = L.DomUtil.create('div', 'heightGraph');
        //controlDiv.title = 'Height Graph Profile';
        this._map = map;
        this._initToggle();
        this._cont = d3.select(controlDiv);
        return controlDiv;
    },
    onRemove: function(map) {
        this._container = null;
    },
    addData: function(data) {
        this._distances = this._calcDistances(data);
        this._heightvalue = this._calculateHeightSteep(data).height;
        this._steepness = this._calculateHeightSteep(data).steep;
        this._dynamicLegend = this._updateLegend(this._steepness);
        this._waypointPosition = this._calculateHeightSteep(data).wpFeatPos;
        this._polygonData = this._updateBarData(this._heightvalue, this._steepness);
        this._waypointData = this._updateWaypointData(this._distances, this._heightvalue);
        this._createSVG = this._createBarChart(this._polygonData, this._waypointData, this._cont, this._heightvalue, this._dynamicLegend);
    },
    _initToggle: function() {
        /* inspired by L.Control.Layers */
        var container = this._container;
        //Makes this work on IE10 Touch devices by stopping it from firing a mouseout event when the touch is released
        container.setAttribute('aria-haspopup', true);
        if (!L.Browser.touch) {
            L.DomEvent.disableClickPropagation(container);
            //.disableScrollPropagation(container);
        } else {
            L.DomEvent.on(container, 'click', L.DomEvent.stopPropagation);
        }
        //this._collapse();
        if (!L.Browser.android) {
            L.DomEvent.on(container, 'click', this._expand, this);
        }
    },
    _expand: function() {
        if (this._container.className.indexOf('heightGraph-collapsed') > -1) {
            this._container.className = this._container.className.replace(' heightGraph-collapsed', '');
        } else {
            L.DomUtil.addClass(this._container, 'heightGraph-collapsed');
        }
    },
    _collapse: function() {
        L.DomUtil.addClass(this._container, 'heightGraph-collapsed');
    },
    /*
     * Reset data
     */
    _clearData: function() {
        this._data = null;
        this._dist = null;
        this._maxElevation = null;
    },
    /*
     * Reset data and display
     */
    clear: function() {
        this._clearData();
        if (!this._areapath) {
            return;
        }
        // workaround for 'Error: Problem parsing d=""' in Webkit when empty data
        // https://groups.google.com/d/msg/d3-js/7rFxpXKXFhI/HzIO_NPeDuMJ
        //this._areapath.datum(this._data).attr("d", this._area);
        this._areapath.attr("d", "M0 0");
        this._x.domain([0, 1]);
        this._y.domain([0, 1]);
        this._updateAxis();
    },
    /**
     * Returns distance between each coordinate of Linestring
     * @param {FeatureCollection} a
     * @returns {Number|Array| Object} Object that contains distances between coordinates, the coordinates, and sum of distances as totaldistance
     */
    _calcDistances: function(a) {
        console.log(a)
        distances = {};
        var first, calc;
        //var wpList=[];
        distances.distance = [];
        distances.wpDistance = [];
        var featureLength = a.features.length;
        distances.coordsOfDist = [];
        for (var i = 0; i < featureLength; i++) {
            var coordLength = a.features[i].geometry.coordinates.length;
            for (var j = 0; j < coordLength - 1; j++) {
                var g = new L.LatLng(a.features[i].geometry.coordinates[j][1], a.features[i].geometry.coordinates[j][0]);
                // catch steps between features
                if (j == 0 && i > 0) {
                    calc = last.distanceTo(g);
                    distances.distance.push(calc);
                    distances.coordsOfDist.push([last, g]);
                }
                var h = new L.LatLng(a.features[i].geometry.coordinates[j + 1][1], a.features[i].geometry.coordinates[j + 1][0]);
                calc = g.distanceTo(h);
                distances.distance.push(calc);
                distances.coordsOfDist.push([g, h]);
                // save last
                if (j == coordLength - 2) {
                    last = h;
                }
            }
        }
        distances.totaldistance = distances.distance.reduce(function(pv, cv) {
            return pv + cv;
        }, 0);
        return distances;
    },
    /**
     * Returns the values (height, steepness) of the FeatureCollection as Array in list
     * @param {FeatureCollection} a
     * @returns {Number|Array|Object} list with height values and steepness values as array
     */
    _calculateHeightSteep: function(a) {
        var height = [];
        var steep = [];
        for (var i = 0; i < a.features.length; i++) {
            var coordNumber = a.features[i].geometry.coordinates.length;
            for (var j = 0; j < coordNumber; j++) {
                height.push(a.features[i].geometry.coordinates[j][2]);
                steep.push(a.features[i].properties.steepness);
            }
        }
        var list = {
            height: height,
            steep: steep
        };
        return list;
    },
    /**
     * Returns values of steepness as Array without duplicates 
     * @param {Array} a: steepness values of all coords
     * @returns {Array} list with steepness color and text as array
     */
    _updateLegend: function(a) {
        var legendList = [];
        //remove duplicates
        var cleanList = a.filter(function(elem, index) {
            return index == a.indexOf(elem);
        });

        function sortNumber(a, b) {
            return a - b;
        }
        cleanList.sort(sortNumber);
        for (var i = 0; i < cleanList.length; i++) {
            var j = cleanList[i] - 5;
            legendList.push(legendData[cleanList[i]][j]);
        }
        return legendList;
    },
    /**
     * Returns list with four x and y coordinates for svg-path (Polygon) and steepness type
     * @param {Array} a: heightvalue
     * @param {Array} steepness
     * @returns {Number|Array|Object} list with coordinates and steepness values as array
     */
    _updateBarData: function(a, steepness) {
        var list = [];
        var wplist = [];
        var adddist = [0];
        var count = a.length;
        for (var i = 0; i < count; i++) {
            adddist[i + 1] = adddist[i] + distances.distance[i];
            list.push({
                coords: [{
                    x: adddist[i],
                    y: a[i]
                }, {
                    x: adddist[(i + 1 == count) ? i : i + 1],
                    y: a[(i + 1 == count) ? i : i + 1]
                }, {
                    x: adddist[(i + 1 == count) ? i : i + 1],
                    y: d3.min(a)
                }, {
                    x: adddist[i],
                    y: d3.min(a)
                }],
                steepness: steepness[i],
                LatLng: distances.coordsOfDist[i]
            });
        }
        return list;
    },
    /**
     * Returns list with two x and y coordinates for svg-path (Line and circle)
     * @param {Array|Object} a: distances
     * @param {Array} b: heightvalue
     * @returns {Number|Array|Object} list with coordinates as array
     */
    _updateWaypointData: function(a, b) {
        var list = [];
        var length = a.wpDistance.length;
        for (var i = 0; i < length; i++) {
            list.push({
                wpCoords: [{
                    x: a.wpDistance[i],
                    y: d3.min(b)
                }, {
                    x: a.wpDistance[i],
                    y: (d3.max(b) / 2)
                }]
            });
        }
        return list;
    },
    _showMarker: function(ll) {
        if (!this._marker) {
            this._marker = new L.circleMarker(ll, {
                radius: 5,
                fillColor: "#ff7800",
                color: "#000",
                weight: 1,
                opacity: 1,
                fillOpacity: 0.8,
            }).addTo(this._map);
        } else {
            this._marker.setLatLng(ll);
        }
    },
    /**
     * @param {Object} polygonData: (x,y-coords, steepness)
     * @param {Object} waypointData: (x,y-coords)
     * @param {Array} heightvalue: heightvalue of each coordinate
     * @param {Object} dynamicLegend: values for legend (text and color)
     */
    _createBarChart: function(polygonData, waypointData, container, heightvalue, dynamicLegend) {
        //SVG area
        var margin = this.options.margins,
            width = this.options.width - margin.left - margin.right,
            height = 180 - margin.top - margin.bottom;
        var x = d3.scale.linear().range([0, width]).domain([0, distances.totaldistance]);
        var y = d3.scale.linear().range([height, 0]).domain(d3.extent(heightvalue, function(d) {
            return d;
        }));
        var xAxis = d3.svg.axis().scale(x).orient("bottom").ticks(4).tickFormat(function(d) {
            return d / 1000;
            // var prefix = d3.formatPrefix(d);
            // return prefix.scale(d) //+ prefix.symbol;
        });
        var yAxis = d3.svg.axis().scale(y).orient("left").ticks(4);
        var tip = d3.tip().attr('class', 'd3-tip').offset([-10, 0]).html(function(d) {
            return ((Math.round(((d.coords[0].y + d.coords[1].y) / 2) * 100) / 100) + " m");
        });
        var tipDist = d3.tip().attr('class', 'd3-tip').offset([-10, 0]).html(function(d) {
            return (d);
        });
        var svgSec = d3.select(this._container).append("svg").attr("class", "background").attr("width", width + margin.left + margin.right).attr("height", height + margin.top + margin.bottom).append("g").attr("transform", "translate(" + margin.left + "," + margin.top + ")");
        // axes and axes labels
        svgSec.append('g').attr("transform", "translate(0," + height + ")") // create a <g> element
            .attr('class', 'x axis') // specify classes
            .call(xAxis).append("text").attr("x", width - 20).attr("y", y(d3.max(heightvalue)) + 15).style("font-size", "12px").style("text-anchor", "middle").attr("fill", "#fff").text("km");
        svgSec.append('g').attr('class', 'y axis').call(yAxis).append("text").attr("x", -30).attr("y", -5).attr("fill", "#fff").attr("dy", "0").style("font-size", "12px").style("text-anchor", "initial").text("hm");
        svgSec.selectAll('.axis path').style({
            'stroke': '#fff',
            'fill': 'none',
            'stroke-width': '1'
        });
        svgSec.selectAll('.axis line').style({
            'visibility': 'hidden'
        });
        // scale data (polygon-path)
        var polygon = d3.svg.line().x(function(d) {
            return x(d.x);
        }).y(function(d) {
            return y(d.y);
        });
        // bar chart as path
        svgSec.selectAll('hpath').data(polygonData).enter().append('path')
            //.attr('leafletId', id)
            .attr('d', function(d) {
                return polygon(d.coords);
            })
            //.attr("data-legend",function(d) { return d.steepness})
            .attr("fill-opacity", 0.6).attr('fill', function(d) {
                return (d.steepness - 5 == -5 ? '#028306' : d.steepness - 5 == -4 ? '#2AA12E' : d.steepness - 5 == -3 ? '#53BF56' : d.steepness - 5 == -2 ? '#7BDD7E' : d.steepness - 5 == -1 ? '#A4FBA6' : d.steepness - 5 == 0 ? '#ffcc99' : d.steepness - 5 == 1 ? '#F29898 ' : d.steepness - 5 == -2 ? '#E07575' : d.steepness - 5 == 3 ? '#CF5352' : d.steepness - 5 == 4 ? '#BE312F' : d.steepness - 5 == 5 ? '#AD0F0C' : '#AD0F0C');
            }).on('mouseover', handleMouseOver).on("mouseout", handleMouseOut).on("mousemove", mousemove);
        // focus line
        var focus = svgSec.append("g").attr("class", "focus").style("display", "none");
        focus.append("text").attr("x", 5).attr("font-size", "10px").attr("dy", ".35em").style("background", "black").attr("fill", "#fff");
        focus.insert("rect", "text").attr("width", function(d) {
            return 53;
        }).attr("height", function(d) {
            return 14;
        }).attr("y", "-7").attr("x", "-1").style("fill", "#000");
        var focusLineGroup = svgSec.append("g").attr("class", "focusLine");
        var focusLine = focusLineGroup.append("line").attr("stroke-width", 1).attr("fill", "#fff").attr("x1", 10).attr("x2", 10).attr("y1", y(d3.max(heightvalue))).attr("y2", y(d3.min(heightvalue))).style("display", "none");
        // waypoints line
        svgSec.selectAll('linePath').data(waypointData).enter().append('path').attr('class', 'wpLine').attr('d', function(d) {
            return polygon(d.wpCoords);
        }).attr('pointer-events', 'none').attr("stroke-width", 2).attr('stroke', 'black');
        // waypoints circle
        svgSec.selectAll('hcircle').data(waypointData).enter().append('circle').attr('cx', function(d) {
                return x(d.wpCoords[1].x);
            }).attr('cy', function(d) {
                return y(d.wpCoords[1].y);
            }).attr('pointer-events', 'none').attr('r', 8)
            //.attr("data-legend",function(d) { return d.steepness})
            .attr('fill', 'black');
        svgSec.selectAll('wpText').data(waypointData).enter().append("text").attr("class", "wp-text").attr("dx", function(d) {
            return x(d.wpCoords[1].x) - 3.5;
        }).attr("dy", function(d) {
            return y(d.wpCoords[1].y - 3);
        }).text(function(d, i) {
            return i + 1;
        }).style('fill', 'white').style('font-size', '12px');
        // legend
        var legendRectSize = 7;
        var legendSpacing = 7;
        var legend = svgSec.selectAll('.legend').data(dynamicLegend).enter().append('g').attr('class', 'legend').attr('transform', function(d, i) {
            var height = legendRectSize + legendSpacing;
            var offset = height * 2;
            var horz = -2 * legendRectSize;
            var vert = i * height - offset;
            return 'translate(' + horz + ',' + vert + ')';
        });
        legend.append('rect').attr('width', legendRectSize).attr('height', legendRectSize).attr('x', 30).attr('y', 30).style('fill', function(d, i) {
            return (d.color);
        });
        legend.append('text').attr('x', 40).attr('y', 36).style('font-size', 10).attr('fill', '#fff').text(function(d, i) {
            return d.text;
        });
        svgSec.call(tip);
        // Create Event Handlers for mouse
        function handleMouseOver(d, i) {
            // Use D3 to select element, change color and size
            tip.show(d);
            focus.style("display", null);
            focusLine.style("display", null);
        }

        function handleMouseOut(d, i) {
            // Use D3 to select element, change color back to normal
            tip.hide(d);
            focus.style("display", "none");
            focusLine.style("display", "none");
        }
        var self = this;

        function mousemove(d) {
            var x0 = x.invert(d3.mouse(this)[0]); //distance in m   
            var d0 = d.coords[0].x,
                d1 = d.coords[1].x;
            var d2 = d1 - x0 > x0 - d0 ? 0 : 1; // shortest distance between mouse and coords of polygon
            var y0 = d.coords[d2].y;
            var LatLngCoords = d.LatLng;
            var segmentCenter = L.latLngBounds(LatLngCoords[0], LatLngCoords[1]).getCenter();
            self._showMarker(segmentCenter);
            focus.style("display", "initial").attr("transform", "translate(" + x(x0) + "," + y(d3.min(heightvalue)) + ")");
            focus.select("text").text(Math.round((x0 / 1000) * 100) / 100 + ' km'); //text in km
            focusLine.style("display", "initial").attr('x1', x(x0)).attr('y1', y(y0)).attr('x2', x(x0));
        }
    }
});
L.control.heightgraph = function(options) {
    return new L.Control.Heightgraph(options);
};